### PR TITLE
Add watson module for processing watson-nlp response to Pandas

### PR DESCRIPTION
**Pull request by BryanCutler**

### CODE REVIEW

File: notebooks/WatsonNLP_Demo.ipynb [hunk 1]
```diff
@@ -0,0 +1,514 @@
+{
```
Comment (frreiss) [inline notebooks/WatsonNLP_Demo.ipynb:1]: <p>Is this string an authentication token?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/25/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='3'/>

Comment (frreiss) [inline notebooks/WatsonNLP_Demo.ipynb:1]: <p>I would leave this uncommented. Showing the length and complexity of the JSON data is an important part of the message of this notebook.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/25/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='5'/>

Comment (frreiss) [inline notebooks/WatsonNLP_Demo.ipynb:1]: <p>Instead of printing out the text representation, the notebook should display each of these dataframes using Jupyter's support for viewing dataframes as HTML tables. One cell/output per dataframe.</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/CODAIT/text-extensions-for-pandas/pull/25/discussion/'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

File: text_extensions_for_pandas/io/test_watson.py [hunk 2]
```diff
@@ -0,0 +1,879 @@
...
+
+from text_extensions_for_pandas.io.watson import *
+
+'''
+text = textwrap.dedent(
+  """\
+  Who is General Failure and why is he reading my hard disk?
+  If Barbie is so popular, why do you have to buy her friends?""")
+'''
+response1 = {
```
Comment (frreiss) [inline text_extensions_for_pandas/io/test_watson.py:27]: You might want to put this data into a JSON file somewhere in the repository instead of an inline Python literal.

File: text_extensions_for_pandas/io/test_watson.py [hunk 3]
```diff
@@ -0,0 +1,879 @@
...
+        # Ensure that diffs are consistent
+        pd.set_option("display.max_columns", 250)
+
+    def tearDown(self):
+        pd.reset_option("display.max_columns")
+
+    def test_print_all(self):
+        result = watson_nlp_parse_response(response1)
+        for k, v in result.items():
+            print("\n\n{}:".format(k))
```
Comment (frreiss) [inline text_extensions_for_pandas/io/test_watson.py:849]: This output should go to a buffer so that it doesn't hide the console output of `unittest` when running regression tests.

File: text_extensions_for_pandas/io/watson.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+#
+# I/O functions related to Watson Natural Language Processing on the IBM Cloud.
+
+import pandas as pd
+import pyarrow as pa
+
+from text_extensions_for_pandas.array import CharSpanArray, TokenSpanArray
+
+
+def watson_nlp_parse_response(response):
```
Comment (frreiss) [inline text_extensions_for_pandas/io/watson.py:27]: This function needs a detailed docstring and a type hint for the return value. What is the type of `response`? What does the function return? What expectations does this function have on the contents of `response` beyond its Python type?

File: text_extensions_for_pandas/io/watson.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+    def flatten_struct(struct_array, parent_name=None):
```
Comment (frreiss) [inline text_extensions_for_pandas/io/watson.py:29]: These inline functions should probably be moved to the top-level scope of this file and renamed to start with an underscore so they don't get picked up by `__all__`. That way the source code of the `watson_nlp_parse_response()` function will come closer to fitting on a single screen.

File: text_extensions_for_pandas/io/watson.py [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+                text_col = pa.concat_arrays(text_col.iterchunks())
+
+                # Flatten to get primitive array convertible to numpy
+                array = location_col.flatten()
+                values = array.to_numpy()
+                begins = values[0::2]
+                ends = values[1::2]
+
+                # Build the covered text, TODO: ok to assume begin is sorted?
+                text = ""
```
Comment (frreiss) [inline text_extensions_for_pandas/io/watson.py:89]: It would be a good idea to add a 1-line check for sortedness (`all(begins[i] <= begins[i+1] for i in range(len(begins)-1))`), as the Watson people aren't terribly good about documenting and/or maintaining these kinds of invariants.

File: text_extensions_for_pandas/io/watson.py [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+                for token, begin in zip(text_col, begins):
+                    if len(text) < begin:
+                        text += " " * (begin - len(text))
+                    text += token.as_py()
+
+                char_span = CharSpanArray(text, begins, ends)
+
+                # TODO: drop location column?
+
+            df = table.to_pandas()
```
Comment (frreiss) [inline text_extensions_for_pandas/io/watson.py:99]: There should be a sanity check here that the schema of the dataframe matches the schema you would expect from the current version of the API, with a clear error message/warning for any deviations.

**Comment by BryanCutler** [comment]
@frreiss , I finished adding spans to the relations DataFrame as we discussed, also did some cleanup, added docs, and some tests. Test coverage is pretty minimal and there are lots of edge cases to cover, so I will continue to add those. I'm removing the draft status and should be ready for review. cc @ZachEichen

File: text_extensions_for_pandas/io/watson.py [hunk 8]
```diff
@@ -0,0 +1,511 @@
...
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#
+
+################################################################################
+# watson.py
+#
+# I/O functions related to Watson Natural Language Understanding on the IBM Cloud.
+# TODO: add links here and docstrings
```
Comment (frreiss) [inline text_extensions_for_pandas/io/watson.py:20]: Can you please add the appropriate links here? Non-IBMers will not know where to look otherwise. Come to think of it, most IBMers wouldn't know either...

File: text_extensions_for_pandas/io/watson.py [hunk 9]
```diff
@@ -0,0 +1,511 @@
...
+import numpy as np
+import pandas as pd
+import pyarrow as pa
+
+from text_extensions_for_pandas.array import CharSpanArray, TokenSpanArray
+from text_extensions_for_pandas.spanner import contain_join
+
+
+# Standard Schemas for Response Data
+entities_schema = [
```
Comment (frreiss) [inline text_extensions_for_pandas/io/watson.py:33]: Unless it was your intent to export these names to user code, I recommend that these constants' names start with an underscore.

File: text_extensions_for_pandas/io/watson.py [hunk 10]
```diff
@@ -0,0 +1,511 @@
...
+    Parse a Watson NLU response as a decoded JSON string, e.g. dictionary containing requested
+    features and convert into a dict of Pandas DataFrames. The following features in the response
+    will be converted:
+        * entities
+        * keywords
+        * relations
+        * semantic_roles
+        * syntax
+
+    TODO: add desc of feature options, link to API
```
Comment (frreiss) [inline text_extensions_for_pandas/io/watson.py:417]: Please add some appropriate links here. This docstring will be part of our Python module's documentation. There should be a link to the main landing page for NLU and to the main documentation page for the APIs.

**Comment by frreiss** [comment]
@BryanCutler thanks for getting this ready. 

The build is currently failing on this branch. There's a lot of unnecessary output in the build log, probably due to the `print()` statements on lines 53-54 of `text_extensions_for_pandas/io/test_watson.py` that haven't yet been fixed.

The actual build failure is occurirng due to this error:
```
======================================================================
ERROR: test_make_bert_tokens (text_extensions_for_pandas.io.test_tokenization.TestTokenize)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/CODAIT/text-extensions-for-pandas/text_extensions_for_pandas/io/test_tokenization.py", line 39, in test_make_bert_tokens
    bert_toks = make_bert_tokens(text, tokenizer)
  File "/home/travis/build/CODAIT/text-extensions-for-pandas/text_extensions_for_pandas/io/tokenization.py", line 52, in make_bert_tokens
    from transformers.tokenization_utils import PreTrainedTokenizerFast
ImportError: cannot import name 'PreTrainedTokenizerFast' from 'transformers.tokenization_utils' (/home/travis/miniconda/envs/pd/lib/python3.7/site-packages/transformers/tokenization_utils.py)
----------------------------------------------------------------------
```
That problem stems from an API change in a library we depend on. I will make a PR shortly to fix that problem.

Can you please fix the extra print statements? Also, none of my review comments from before are marked "resolved". Can you please make sure that each of the things I flagged has been dealt with and mark conversations as resolved accordingly?

**Comment by frreiss** [comment]
Opened #42 to fix the build break.

File: notebooks/Person.ipynb [hunk 11]
```diff
@@ -520,7 +520,7 @@
    "name": "python",
    "nbconvert_exporter": "python",
    "pygments_lexer": "ipython3",
-   "version": "3.7.6"
+   "version": "3.7.7"
```
Comment (BryanCutler) [inline notebooks/Person.ipynb:523]: oops, I'll revert this

**Comment by frreiss** [comment]
LGTM.

**Comment by BryanCutler** [comment]
Thanks @frreiss !
